utils/tls/tls.go

```diff
@@ -0,0 +1,33 @@
+package tls
+
+import (
+	"crypto/tls"
+	"time"
+)
+
+type CachedCertificateLoader struct {
+	CertPath    string
+	KeyPath     string
+	certificate *tls.Certificate
+	nextReload  time.Time
+}
+
+// Get Certificate uses the cachedCertificateLoader struct
+// to automatically reload the certificate from disk every hour
+func (c *CachedCertificateLoader) GetCertificate(nowFunc func() time.Time) (*tls.Certificate, error) {
+	var now time.Time
+	if nowFunc == nil {
+		now = time.Now()
+	} else {
+		now = nowFunc()
+	}
+	if now.After(c.nextReload) {
+		cert, err := tls.LoadX509KeyPair(c.CertPath, c.KeyPath)
+		if err != nil {
+			return nil, err
+		}
+		c.certificate = &cert
```

Comment:
Don't you need to protect this with a lock?

Reply:
Hmm. I copied it from our golang mtls code, I assume it is production ready.

In golang writing to maps is not lock-safe, but just setting a variable is right?

Reply:
> In golang writing to maps is not lock-safe, but just setting a variable is right?

Yea, that's the case as I understand it. So there's a race here where _n_ goroutines getting a cert at the same time will all have their gorountines reading from disk, but AFAIK this just introduces extra overhead but wouldn't be buggy. You could do some locking if you want to avoid that.

Reply:
Unless the TLS library is doing locking on our behalf, I'm not sure it is. There's a very clear race condition though, where if two routines hit this at the same time, they will both do a read-modify-write on the variable, and that will trigger the race detector.

Reply:
Ok, I think I've added a mutex to lock the loading/setting of the cert, but only if it is time to do so.